src/provider/controller/IntentHandler.ts

```diff
@@ -118,42 +118,44 @@ export class IntentHandler {
 
     private async fireIntent(intent: Intent, appInfo: Application): Promise<IntentResolution> {
         await this._model.ensureRunning(appInfo);
-
-        // TODO: Revisit timeout logic [SERVICE-556]
-        let dispatchingCompleted = false;
-        const dispatchResults = await withTimeout(Timeouts.ADD_INTENT_LISTENER, (async () => {
-            const appWindows = await this._model.expectWindowsForApp(appInfo);
-
-            // Wait for windows to add intent listener, then dispatch payload
-            return Promise.all(appWindows.map(async (window: AppWindow): Promise<boolean> => {
-                if (await window.isReadyToReceiveIntent(intent.type)) {
-                    const payload: ReceiveIntentPayload = {context: intent.context, intent: intent.type};
-
-                    // TODO: Implement a timeout so a misbehaving intent handler can't block the intent raiser [SERVICE-555]
-                    if (!dispatchingCompleted) {
-                        await this._apiHandler.dispatch(window.identity, APIToClientTopic.RECEIVE_INTENT, payload);
-                        return true;
-                    }
-                }
-                return false;
-            }));
-        })());
-
-        dispatchingCompleted = true;
-
-        if (dispatchResults[0] || !dispatchResults[1]!.includes(true)) {
-            throw new FDC3Error(ResolveError.IntentTimeout, `Timeout waiting for intent listener to be added for intent: ${intent.type}`);
+        const appWindows = await this._model.expectWindowsForApp(appInfo);
+        const promises = appWindows.map(async (window: AppWindow): Promise<any> => {
+            if (await window.isReadyToReceiveIntent(intent.type)) {
+                const payload: ReceiveIntentPayload = {context: intent.context, intent: intent.type};
+                return this._apiHandler.dispatch(window.identity, APIToClientTopic.RECEIVE_INTENT, payload);
+            } else {
+                throw new Error(`${appInfo.name} not ready to recieve intents.`);
+            }
+        });
+
+        let data: unknown;
+        try {
+            // Use the first handler return data that matches the predicate
+            const [didTimeout, result] = await withTimeout(Timeouts.INTENT_RESOLUTION, raceTilPredicate(promises, returnValuePredicate));
+            if (didTimeout) {
+                throw new Error('Race timed out');
+            }
+            data = result;
+        } catch (error) {
+            if (/Exceptions in all/.test(error.message)) {
```

Comment:
Is there an alternative to matching against error messages here? Feels fragile

Reply:
Reworked the flow of this.